File: src/main/java/com/netsensia/rivalchess/bitboards/EngineBitboards.java
package com.netsensia.rivalchess.bitboards;

import com.netsensia.rivalchess.model.Colour;
import com.netsensia.rivalchess.model.SquareOccupant;

import java.util.Arrays;

import static com.netsensia.rivalchess.bitboards.util.BitboardUtilsKt.getFirstOccupiedSquare;
import static com.netsensia.rivalchess.bitboards.util.BitboardUtilsKt.isBishopAttackingSquare;
import static com.netsensia.rivalchess.bitboards.util.BitboardUtilsKt.isRookAttackingSquare;
import static com.netsensia.rivalchess.bitboards.util.BitboardUtilsKt.getPawnMovesCaptureOfColour;
import static com.netsensia.rivalchess.bitboards.BitboardConstantsKt.*;

public class EngineBitboards {

    private static EngineBitboards INSTANCE;

    private long[] pieceBitboards;

    public static EngineBitboards getInstance() {
        if(INSTANCE == null) {
            INSTANCE = new EngineBitboards();
        }

        return INSTANCE;
    }

    private EngineBitboards () {
        reset();
    }

    public long getAllPieceBitboard() {
        return getPieceBitboard(BitboardType.ALL);
    }

    public void xorPieceBitboard(int i, long xorBy) {
        this.pieceBitboards[i] ^= xorBy;
    }

    public void xorPieceBitboard(BitboardType type, long xorBy) {
        this.pieceBitboards[type.getIndex()] ^= xorBy;
    }

    public void orPieceBitboard(BitboardType type, long xorBy) {
        this.pieceBitboards[type.getIndex()] |= xorBy;
    }

    public void reset() {
        pieceBitboards = new long[BitboardType.getNumBitboardTypes()];
        Arrays.fill(pieceBitboards, 0);
    }

    public void setPieceBitboard(BitboardType type, long bitboard) {
        pieceBitboards[type.getIndex()] = bitboard;
    }

    public long getPieceBitboard(BitboardType type) {
        return pieceBitboards[type.getIndex()];
    }

    public void movePiece(SquareOccupant piece, int compactMove) {

        final byte moveFrom = (byte) (compactMove >>> 16);
        final byte moveTo = (byte) (compactMove & 63);

        final long fromMask = 1L << moveFrom;
        final long toMask = 1L << moveTo;

        pieceBitboards[piece.getIndex()] ^= fromMask | toMask;
    }

    public long getRookMovePiecesBitboard(Colour colour) {
        return colour == Colour.WHITE
                ? getPieceBitboard(BitboardType.WR) | getPieceBitboard(BitboardType.WQ)
                : getPieceBitboard(BitboardType.BR) | getPieceBitboard(BitboardType.BQ);
    }

    public long getBishopMovePiecesBitboard(Colour colour) {
        return colour == Colour.WHITE
                ? getPieceBitboard(BitboardType.WB) | getPieceBitboard(BitboardType.WQ)
                : getPieceBitboard(BitboardType.BB) | getPieceBitboard(BitboardType.BQ);
    }

    public boolean isSquareAttackedBy(final int attackedSquare, final Colour attacker) {

        if ((pieceBitboards[SquareOccupant.WN.ofColour(attacker).getIndex()] & getKnightMoves().get(attackedSquare)) != 0 ||
                (pieceBitboards[SquareOccupant.WK.ofColour(attacker).getIndex()] & getKingMoves().get(attackedSquare)) != 0 ||
                (pieceBitboards[SquareOccupant.WP.ofColour(attacker).getIndex()]
                        & getPawnMovesCaptureOfColour(attacker.opponent()).get(attackedSquare)) != 0)
            return true;

        long bitboardBishop = getBishopMovePiecesBitboard(attacker);

        while (bitboardBishop != 0) {
            final int pieceSquare = getFirstOccupiedSquare(bitboardBishop);
            bitboardBishop ^= (1L << (pieceSquare));
            if (isBishopAttackingSquare(attackedSquare, pieceSquare, getAllPieceBitboard())) {
                return true;
            }
        }

        long bitboardRook = getRookMovePiecesBitboard(attacker);

        while (bitboardRook != 0) {
            final int pieceSquare = Long.numberOfTrailingZeros(bitboardRook);
            bitboardRook ^= (1L << (pieceSquare));
            if (isRookAttackingSquare(attackedSquare, pieceSquare, getAllPieceBitboard())) {
                return true;
            }
        }

        return false;
    }
}

File: src/main/java/com/netsensia/rivalchess/bitboards/EngineBitboards.kt
package com.netsensia.rivalchess.bitboards

import com.netsensia.rivalchess.bitboards.util.*
import com.netsensia.rivalchess.consts.*
import com.netsensia.rivalchess.model.Colour
import java.util.*
import kotlin.collections.copyOf

class EngineBitboards() {
    @JvmField
    var pieceBitboards = LongArray(BITBOARD_COUNT)

    init { reset() }

    constructor(thoseBitboards: EngineBitboards) : this() {
        pieceBitboards = thoseBitboards.pieceBitboards.copyOf()
    }

    fun reset() {
        pieceBitboards = longArrayOf(0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0)
    }

    fun xorPieceBitboard(i: Int, xorBy: Long) {
        pieceBitboards[i] = pieceBitboards[i] xor xorBy
    }

    fun orPieceBitboard(type: Int, xorBy: Long) {
        pieceBitboards[type] = pieceBitboards[type] or xorBy
    }

    fun setPieceBitboard(type: Int, bitboard: Long) {
        pieceBitboards[type] = bitboard
    }

    fun getWhitePieces() =
            pieceBitboards[BITBOARD_WK] or
            pieceBitboards[BITBOARD_WN] or
            pieceBitboards[BITBOARD_WQ] or
            pieceBitboards[BITBOARD_WB] or
            pieceBitboards[BITBOARD_WR] or
            pieceBitboards[BITBOARD_WP]

    fun getBlackPieces() =
            pieceBitboards[BITBOARD_BK] or
            pieceBitboards[BITBOARD_BN] or
            pieceBitboards[BITBOARD_BQ] or
            pieceBitboards[BITBOARD_BB] or
            pieceBitboards[BITBOARD_BR] or
            pieceBitboards[BITBOARD_BP]

    fun movePiece(piece: Int, compactMove: Int) {
        val fromMask = (1L shl (compactMove ushr 16))
        val toMask = (1L shl (compactMove and 63))
        xorPieceBitboard(piece, (fromMask or toMask))
    }

    private fun getRookMovePiecesBitboard(colour: Colour) =
        if (colour == Colour.WHITE) pieceBitboards[BITBOARD_WR] or pieceBitboards[BITBOARD_WQ] else
            pieceBitboards[BITBOARD_BR] or pieceBitboards[BITBOARD_BQ]

    private fun getBishopMovePiecesBitboard(colour: Colour) =
        if (colour == Colour.WHITE) pieceBitboards[BITBOARD_WB] or pieceBitboards[BITBOARD_WQ] else
            pieceBitboards[BITBOARD_BB] or pieceBitboards[BITBOARD_BQ]

    fun isSquareAttackedBy(attackedSquare: Int, attacker: Colour): Boolean {

        if (pieceBitboards[if (attacker == Colour.WHITE) BITBOARD_WN else BITBOARD_BN] and knightMoves[attackedSquare] != 0L ||
                pieceBitboards[if (attacker == Colour.WHITE) BITBOARD_WK else BITBOARD_BK] and kingMoves[attackedSquare] != 0L ||
                (pieceBitboards[if (attacker == Colour.WHITE) BITBOARD_WP else BITBOARD_BP] and
                        getPawnMovesCaptureOfColour(attacker.opponent())[attackedSquare]) != 0L) return true

        applyToSquares(getBishopMovePiecesBitboard(attacker)) {
            if (isBishopAttackingSquare(attackedSquare, it, pieceBitboards[BITBOARD_ALL])) return true
        }

        applyToSquares(getRookMovePiecesBitboard(attacker)) {
            if (isRookAttackingSquare(attackedSquare, it, pieceBitboards[BITBOARD_ALL])) return true
        }

        return false
    }
}
